### PR TITLE
PR: Unify UI of editors in Variable Explorer and simplify code

### LINKED ISSUE #21556: Follow up to work that added refresh buttons to editors from the Variable Explorer
This is a series of tasks that I think we should address after merging PR #21312:

- Make editors inherit from `SpyderWidgetMixin` and use its API to build actions, menus and toolbars. I think that's important because those widgets are quite complex now, so if we plan to add more functionality to them in the future, it'd be nice if we could do it in a more structured way. And using that API could help with that.
- Review the Dataframe editor toolbar and context menu ordering. I think that toolbar could see a bit love to look a bit better. We can talk about it here.
- Remove checkboxes in Dataframe editor and add those as buttons to its toolbar.
- Fix margin and alignment of buttons in all editors. Specifically:
  - There's too much margin between the content area and the buttons.
  - Buttons are not aligned to the right of the content area, which makes the layout look odd.
- Make Object explorer toolbar look similar to the toolbars shown in the other editors.

### PR DESCRIPTION
**Pull request by jitseniesen**
## Description of Changes

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

This PR is a follow-up to PR #21312 which added the Refresh button, in order to make the UI and the code in the array editor, collections editor, dataframe editor and object explorer more uniform, as suggested in issue #21556:
* Redistribute and reorder the actions in the context menus and the toolbars so that the UI is more logical and consistent.
* Move buttons in bottom left corner of dataframe editor (Format, Resize, Background Color, Column Min/Max) to toolbar.
* Tweak the UI for consistency: no vertical space between toolbar and main widget, and standardize vertical space between main widget and "Close" button at the bottom and align right edge.
* Use standard functions from SpyderWidgetMixin to create actions, menus and toolbars; this changes the appearance of the toolbar in the object explorer.
* Minor refactoring to simplify the code.

After the PR, the editors look as follows:

followup-array-new3
followup-coll-new3
followup-df-new3
followup-obj-new2
followup-varexp-new3

Here are the corresponding images from before the PR:

followup-array-old
followup-coll-old
followup-df-old
followup-obj-old
followup-varexp-old

### Issue(s) Resolved

Fixes #21556

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: Jitse Niesen

### CODE REVIEW

**Comment by jitseniesen** [comment]
I added a commit which reflects the latest comment on the issue regarding the UI. I also updated the screenshots at the top of this pull request and added a screenshot of the Variable Editor itself.

**Review by CAM-Gerlach — COMMENTED** [review]
Thanks so much for these great improvements, @jitseniesen ! (Didn't see this until now, sorry,  since I was watching the issue but not this PR)

I just have a few followup comments/suggestions:

* [x] Seems like `insert above` is _below_ `insert below` for some reason? Seems like they should be reversed (as in my original suggestion)
* [x] Seems like the refresh button is on the left rather than the right in the collections editor toolbar?
* [x] The (normally blank) circular progress indicator leaves a gap between the refresh button and the options menu; we should move that to the left of the right toolbar button group (i.e. to the left of the search button)
* [x] Upon seeing how they look in the UI, I think we should split the row and column actions in DF viewer context menu and toolbar into separate groups, instead of having them all in one big one. This would make them easier to navigate and not seem too overwhelming to users. So instead of `Insert above`, `Insert below`, `Insert before` `Insert after` `Duplicate row` `Duplicate column` `Remove Row` `Remove Column`; I suggest `Insert above`, `Insert below`, `Duplicate row` `Remove Row` `|`  `Insert before` `Insert after` `Duplicate column` `Remove Column`.
* [ ] Finally, I went back and forth on this a lot before and almost suggested it previously but didn't, but given the column resize buttons in the array and df editors affect only the presentation rather than the content itself, I think we should move them to immediately to the left of the refresh button instead of as a separate group at the beginning of the toolbar. This also puts core content-focused toolbar actions first in the df editor, as is generally expected and done in the other toolbars.

Thanks!

**Comment by jitseniesen** [comment]
Thanks @CAM-Gerlach for your further comments. I implemented them and updated the screenshots at the top of this PR.

> Seems like `insert above` is _below_ `insert below` for some reason? Seems like they should be reversed (as in my original suggestion)

Done, under the assumption that you meant that `insert after` and `insert before` should be reversed.

> Seems like the refresh button is on the left rather than the right in the collections editor toolbar?

Done.

> The (normally blank) circular progress indicator leaves a gap between the refresh button and the options menu; we should move that to the left of the right toolbar button group (i.e. to the left of the search button)

Yes, I noticed that as well and I was planning to ask you about it, but I forgot, so I'm glad you picked that up. I did not find a clear way to do this using Spyder's API so I ended up overriding the private `_setup()` method of `PluginMainWidget`.

> Upon seeing how they look in the UI, I think we should split the row and column actions in DF viewer context menu and toolbar into separate groups, instead of having them all in one big one. This would make them easier to navigate and not seem too overwhelming to users. So instead of `Insert above`, `Insert below`, `Insert before` `Insert after` `Duplicate row` `Duplicate column` `Remove Row` `Remove Column`; I suggest `Insert above`, `Insert below`, `Duplicate row` `Remove Row` `|`  `Insert before` `Insert after` `Duplicate column` `Remove Column`.

Done. 

> Finally, I went back and forth on this a lot before and almost suggested it previously but didn't, but given the column resize buttons in the array and df editors affect only the presentation rather than the content itself, I think we should move them to immediately to the left of the refresh button instead of as a separate group at the beginning of the toolbar. This also puts core content-focused toolbar actions first in the df editor, as is generally expected and done in the other toolbars.

Done. This does mean that the toolbar in the array editor only has buttons on the right and nothing on the left, but so be it. I did the same move in the collection editor.

File: spyder/plugins/variableexplorer/widgets/dataframeeditor.py [hunk 1]
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
...
-                    self.toolbar.addAction(item)
+        self.toolbar.addAction(self.dataTable.insert_action_above)
+        self.toolbar.addAction(self.dataTable.insert_action_below)
+        self.toolbar.addAction(self.dataTable.duplicate_row_action)
+        self.toolbar.addAction(self.dataTable.remove_row_action)
+        self.toolbar.addSeparator()
+        self.toolbar.addAction(self.dataTable.insert_action_before)
+        self.toolbar.addAction(self.dataTable.insert_action_after)
+        self.toolbar.addAction(self.dataTable.duplicate_col_action)
+        self.toolbar.addAction(self.dataTable.remove_col_action)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/dataframeeditor.py:1819]: Is it not possible to use `add_item_to_toolbar` (which support sections) here?

File: spyder/plugins/variableexplorer/widgets/dataframeeditor.py [hunk 2]
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
...
+
+        stretcher = self.create_stretcher('Toolbar stretcher')
+        self.toolbar.addWidget(stretcher)
+
+        self.toolbar.addAction(self.dataTable.resize_action)
+        self.toolbar.addAction(self.dataTable.resize_columns_action)
+        self.toolbar.addAction(self.refresh_action)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/dataframeeditor.py:1826]: And finally you'd need to add this actions (I'm not sure if you need to declare a section for them or not).

File: spyder/plugins/variableexplorer/widgets/dataframeeditor.py [hunk 3]
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
...
+
+        options_menu = self.create_menu('Options menu', register=False)
+        options_menu.add_action(self.bgcolor_action)
+        options_menu.add_action(self.bgcolor_global_action)
+        options_menu.add_action(self.format_action)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/dataframeeditor.py:1831]:
```suggestion
        for action in [self.bgcolor_action, self.bgcolor_global_action,
                       self.format_action]:
            self.add_item_to_menu(action, options_menu)

```

File: spyder/plugins/variableexplorer/widgets/dataframeeditor.py [hunk 4]
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
...
+        options_button = self.create_toolbutton(
+            name='Options toolbutton',
+            text=_('Options'),
+            icon=ima.icon('tooloptions'),
+            register=False
+        )
+        options_button.setPopupMode(QToolButton.InstantPopup)
+        options_button.setMenu(options_menu)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/dataframeeditor.py:1839]:
```suggestion
        options_button.setMenu(options_menu)

```

File: spyder/plugins/variableexplorer/widgets/dataframeeditor.py [hunk 5]
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
...
+        self.toolbar.addWidget(options_button)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/dataframeeditor.py:1840]: This should also use the toolbar API.

File: spyder/plugins/variableexplorer/widgets/dataframeeditor.py [hunk 6]
```diff
@@ -1790,14 +1850,15 @@ def save_and_close_enable(self, top_left, bottom_right):
...
+            text=_("Edit"),
             icon=ima.icon('edit'),
-            triggered=lambda: self.edit_header_item(header=header)
+            triggered=lambda: self.edit_header_item(header=header),
+            register_action=False
         )
         header_menu = [edit_header_action]
-        menu = SpyderMenu(self)
+        menu = self.create_menu('Context header menu', register=False)
         add_actions(menu, header_menu)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/dataframeeditor.py:1862]:
```suggestion
        self.add_item_to_menu(menu, header_menu)
```

File: spyder/plugins/variableexplorer/widgets/main_widget.py [hunk 7]
```diff
@@ -435,6 +436,26 @@ def setup(self):
...
+        super()._setup()
+
+        corner_widget = self._corner_widget
+        for action in corner_widget.actions():
+            if action.defaultWidget() == self.get_options_menu_button():
+                options_menu_action = action
+
+        corner_widget.insertAction(options_menu_action, self.search_action)
+        corner_widget.insertAction(options_menu_action, self.filter_button)
+        corner_widget.insertAction(options_menu_action, self.refresh_action)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/main_widget.py:457]:
```suggestion
        for action in [self.search_action, self.filter_button,
                       self.refresh_action]:
            corner_widget.insertAction(options_menu_action, action)
```

File: spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py [hunk 8]
```diff
@@ -15,26 +15,23 @@
...
-from spyder.config.manager import CONF
 from spyder.plugins.variableexplorer.widgets.basedialog import BaseDialog
 from spyder.plugins.variableexplorer.widgets.objectexplorer import (
     DEFAULT_ATTR_COLS, DEFAULT_ATTR_DETAILS, ToggleColumnTreeView,
     TreeItem, TreeModel, TreeProxyModel)
 from spyder.utils.icon_manager import ima
-from spyder.utils.qthelpers import (
-    add_actions, create_toolbutton, qapplication, safe_disconnect)
-from spyder.utils.stylesheet import PANES_TOOLBAR_STYLESHEET
+from spyder.utils.qthelpers import add_actions, qapplication
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:33]:
```suggestion
from spyder.utils.qthelpers import qapplication
```

As before, `add_actions` shouldn't be necessary, so please replace its usage with the appropriate API method.

File: spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py [hunk 9]
```diff
@@ -205,122 +179,125 @@ def set_value(self, obj):
...
+            name='Show callable attributes',
+            text=_("Show callable attributes"),
+            icon=ima.icon("class"),
+            tip=_("Shows/hides attributes that are callable "
+                  "(functions, methods etc)"),
+            toggled=self._show_callable_attributes,
+            option='show_callable_attributes',
+            register_action=False
         )
+        self.toolbar.add_item(self.toggle_show_callable_action)
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:218]: I think it be better if you introduce all items after creating them first (i.e. at the end of this method) and by using `add_item_to_toolbar`.

File: spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py [hunk 10]
```diff
@@ -205,122 +179,125 @@ def set_value(self, obj):
...
         bottom_layout = QHBoxLayout()
         bottom_layout.setSpacing(0)
-        bottom_layout.setContentsMargins(5, 5, 5, 5)  # left top right bottom
+        bottom_layout.setContentsMargins(0, 0, 0, 0)
         bottom_pane_widget.setLayout(bottom_layout)
         self.central_splitter.addWidget(bottom_pane_widget)
 
         group_box = QGroupBox(_("Details"))
+        group_box.setStyleSheet('QGroupBox '
+                                '{margin-bottom: 0px; margin-right: -2px;}')
```
Comment (ccordoba12) [inline spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:294]:
```suggestion
        group_box.setStyleSheet(
            'QGroupBox {margin-bottom: 0px; margin-right: -2px;}'
        )
```

File: spyder/widgets/collectionseditor.py [hunk 11]
```diff
@@ -45,24 +45,21 @@
...
-from spyder.api.widgets.toolbars import SpyderToolbar
+from spyder.api.widgets.mixins import SpyderWidgetMixin
 from spyder.config.base import _, running_under_pytest
 from spyder.py3compat import (is_binary_string, to_text_string,
                               is_type_text_string)
 from spyder.utils.icon_manager import ima
 from spyder.utils.misc import getcwd_or_home
-from spyder.utils.qthelpers import (
-    add_actions, create_action, MENU_SEPARATOR, mimedata2url)
+from spyder.utils.qthelpers import add_actions, MENU_SEPARATOR, mimedata2url
```
Comment (ccordoba12) [inline spyder/widgets/collectionseditor.py:54]:
```suggestion
from spyder.utils.qthelpers import mimedata2url
```

File: spyder/widgets/collectionseditor.py [hunk 12]
```diff
@@ -644,99 +642,146 @@ def setup_table(self):
...
             self.plot_action,
             self.hist_action,
-            self.imshow_action,
-            MENU_SEPARATOR,
-            resize_action,
-            resize_columns_action
+            self.imshow_action
         ]
-        add_actions(menu, self.menu_actions)
+        add_actions(menu, menu_actions)
```
Comment (ccordoba12) [inline spyder/widgets/collectionseditor.py:782]: This menu should also be built as explained above, i.e. by declaring sections and then using `add_item_to_menu`.

File: spyder/widgets/collectionseditor.py [hunk 13]
```diff
@@ -644,99 +642,146 @@ def setup_table(self):
...
 
-        self.empty_ws_menu = SpyderMenu(self)
+        self.empty_ws_menu = self.create_menu('Empty ws', register=False)
         add_actions(
```
Comment (ccordoba12) [inline spyder/widgets/collectionseditor.py:785]: Please replace this usage of `add_actions` (I can't leave a suggestion for it).

File: spyder/widgets/collectionseditor.py [hunk 14]
```diff
@@ -1464,27 +1509,38 @@ def __init__(self, parent, data, namespacebrowser=None,
...
+            name=None,
             text=_('Refresh'),
             icon=ima.icon('refresh'),
             tip=_('Refresh editor with current value of variable in console'),
-            triggered=lambda: self.sig_refresh_requested.emit()
+            triggered=lambda: self.sig_refresh_requested.emit(),
+            register_action=None
         )
+
+        toolbar = self.create_toolbar('Editor toolbar', register=False)
```
Comment (ccordoba12) [inline spyder/widgets/collectionseditor.py:1521]: If possible, please build this toolbar with our API.

**Comment by jitseniesen** [comment]
@ccordoba12 As requested, the PR is now using add_item_to_menu() and add_item_to_toolbar() from SpyderWidgetMixin to build the menus and toolbars, and it's using sections to add the separators. For consistency, I also "declared" all the names of actions and widgets in enum-style classes like DataframeEditorActions. It does make the code more verbose but I guess that is the price to pay for consistency. I also addressed your other comments.

The PR does need a rebase, but I don't want to mess up your review process. So please let me know when you want to rebase.

**Review by CAM-Gerlach — APPROVED** [review]
LGTM now from my side; thanks @jitseniesen !

**Review by ccordoba12 — APPROVED** [review]
This is a terrific improvement!! Thanks @jitseniesen for your hard work on this!

**Comment by ccordoba12** [comment]
Note: There are some really minor issues about this that I'd like to discuss with @jitseniesen and @CAM-Gerlach, but I'm going to merge because @dalthviz needs this so he can finish the migration of the editor to the new API.

So, I'll open a new issue about them.

**Pull request merged**